lib/Target/OpenFhePke/OpenFhePkeEmitter.cpp

```diff
@@ -65,6 +70,88 @@ FailureOr<std::string> getStringForConstant(Value value) {
   return failure();
 }
 
+// Returns true if the given array of sizes is contiguous.
+bool isSingleContiguousSlice(ArrayRef<int64_t> sizes,
+                             ArrayRef<int64_t> sourceShape) {
+  // Find and drop leading and trailing unit sizes.
+  ArrayRef<int64_t> range = sizes;
+  const auto *nonUnitFrontIt =
+      llvm::find_if(range, [](int64_t size) { return size != 1; });
+  size_t numUnitFrontSizes = std::distance(range.begin(), nonUnitFrontIt);
+  int64_t numDroppedFrontDims = std::min(numUnitFrontSizes + 1, range.size());
+  const auto nonUnitBackIt = llvm::find_if(
+      llvm::reverse(range), [](int64_t size) { return size != 1; });
+  size_t numUnitBackSizes =
+      std::distance(llvm::reverse(range).begin(), nonUnitBackIt);
+  int64_t numDroppedBackDims = std::min(numUnitBackSizes, range.size());
+
+  auto drop = [numDroppedFrontDims, numDroppedBackDims](auto range) {
+    return range.drop_front(numDroppedFrontDims).drop_back(numDroppedBackDims);
+  };
+
+  // Check that the remaining size matches the source shape.
+  return drop(range) == drop(sourceShape);
+}
+
+/// Print the integer element of a DenseElementsAttr.
+static void printDenseIntElement(const APInt &value, raw_ostream &os,
+                                 Type type) {
+  if (type.isInteger(1))
+    os << (value.getBoolValue() ? "true" : "false");
+  else
+    value.print(os, !type.isUnsignedInteger());
+}
+
+/// Print a floating point value in a way that the parser will be able to
+/// round-trip losslessly.
+static void printFloatValue(const APFloat &apValue, raw_ostream &os) {
+  assert(apValue.isFinite() && "expected finite value");
+  SmallString<128> strValue;
+  apValue.toString(strValue, /*FormatPrecision=*/6, /*FormatMaxPadding=*/0,
+                   /*TruncateZero=*/false);
+  // Parse back the stringized version and check that the value is equal
+  // (i.e., there is no precision loss). If it is not, use the default format of
+  // APFloat instead of the exponential notation.
+  if (!APFloat(apValue.getSemantics(), strValue).bitwiseIsEqual(apValue)) {
```

Comment:
Devious. I love it.

Reply:
It does raise the question to me: why is the default not good enough always? It's not quite explained here.

Reply:
it was only to prefer some readable format, but the default will always work!